src/openssl/digest.c

```diff
@@ -94,10 +143,27 @@ bool digest_create(digest_t *digest, const void *indata, size_t inlen, void *out
 	unsigned char tmpdata[len];
 
 	if(digest->hmac_ctx) {
-		if(!HMAC_Init_ex(digest->hmac_ctx, NULL, 0, NULL, NULL)
-		                || !HMAC_Update(digest->hmac_ctx, indata, inlen)
-		                || !HMAC_Final(digest->hmac_ctx, tmpdata, NULL)) {
-			logger(DEBUG_ALWAYS, LOG_DEBUG, "Error creating digest: %s", ERR_error_string(ERR_get_error(), NULL));
+		bool ok;
+
+#if OPENSSL_VERSION_MAJOR < 3
+		ok = HMAC_Init_ex(digest->hmac_ctx, NULL, 0, NULL, NULL)
+		     && HMAC_Update(digest->hmac_ctx, indata, inlen)
+		     && HMAC_Final(digest->hmac_ctx, tmpdata, NULL);
+#else
+		// The cast is okay, the parameter is not going change. For example, see:
+		// https://github.com/openssl/openssl/blob/31b7f23d2f958491d46c8a8e61c2b77b1b546f3e/crypto/ec/ecdh_kdf.c#L37-L38
+		const OSSL_PARAM params[] = {
+			OSSL_PARAM_utf8_string(OSSL_STORE_PARAM_DIGEST, (char *)digest->hmac_algo, 0),
+			OSSL_PARAM_END,
+		};
+
+		ok = EVP_MAC_init(digest->hmac_ctx, digest->hmac_key, digest->hmac_key_len, params)
+		     && EVP_MAC_update(digest->hmac_ctx, indata, inlen)
+		     && EVP_MAC_final(digest->hmac_ctx, tmpdata, NULL, sizeof(tmpdata));
```

Comment:
Initializing a HMAC context is slow. That's why we want to call it with NULL parameters, otherwise you'd just use the one-shot functions, like `EVP_Q_mac()`. It looks like you should still be able to do pass NULL key and params if you set the key beforehand using `EVP_MAC_CTX_set_params()`, although I do wonder if it still caches all the state derived from the key...